oem/ibm/libpldmresponder/collect_slot_vpd.cpp

```diff
@@ -0,0 +1,340 @@
+         "SLOT_OBJ", slotObjectPath);
+
+    std::string adapterObjPath;
+    try
+    {
+        // get the adapter dbus object path from the slot dbus object path
+        adapterObjPath = getAdapterObjPath(slotObjectPath).value();
+    }
+    catch (const std::bad_optional_access& e)
+    {
+        error("Failed to get the adapter dbus object ERROR={ERROR}", "ERROR",
+              e);
+        return;
+    }
+
+    info(
+        "CM: Found an adapter under the slot, adapter object:{ADAPTER_OBJ_PATH}",
+        "ADAPTER_OBJ_PATH", adapterObjPath);
+    // create a presence match for the adpter present property
+    createPresenceMatch(adapterObjPath, entity, stateFiledValue);
+
+    // call the VPD Manager to collect/remove VPD objects
+    callVPDManager(adapterObjPath, stateFiledValue);
+
+    // start the 1 min timer
+    timer.restart(std::chrono::seconds(60));
+}
+
+void SlotHandler::callVPDManager(const std::string& adapterObjPath,
+                                 uint8_t stateFiledValue)
+{
+    static constexpr auto VPDObjPath = "/com/ibm/VPD/Manager";
+    static constexpr auto VPDInterface = "com.ibm.VPD.Manager";
+    auto& bus = pldm::utils::DBusHandler::getBus();
+    try
+    {
+        auto service = pldm::utils::DBusHandler().getService(VPDObjPath,
+                                                             VPDInterface);
+        if (stateFiledValue == uint8_t(PLDM_OEM_IBM_PCIE_SLOT_EFFECTER_ADD))
+        {
+            auto method = bus.new_method_call(service.c_str(), VPDObjPath,
+                                              VPDInterface, "CollectFRUVPD");
+            method.append(
+                static_cast<sdbusplus::message::object_path>(adapterObjPath));
+            bus.call_noreply(method, dbusTimeout);
+        }
+        else if (stateFiledValue ==
+                     uint8_t(PLDM_OEM_IBM_PCIE_SLOT_EFFECTER_REMOVE) ||
+                 stateFiledValue ==
+                     uint8_t(PLDM_OEM_IBM_PCIE_SLOT_EFFECTER_REPLACE))
+        {
+            auto method = bus.new_method_call(service.c_str(), VPDObjPath,
+                                              VPDInterface, "deleteFRUVPD");
+            method.append(
+                static_cast<sdbusplus::message::object_path>(adapterObjPath));
+            bus.call_noreply(method, dbusTimeout);
+        }
+    }
+    catch (const std::exception& e)
+    {
+        error(
+            "failed to make a d-bus call to VPD Manager , Operation = {STATE_FILED_VAL}, ERROR={ERROR}",
+            "STATE_FILED_VAL", (unsigned)stateFiledValue, "ERROR", e);
+    }
+}
+
+std::optional<std::string>
+    SlotHandler::getAdapterObjPath(const std::string& slotObjPath)
+{
+    static constexpr auto searchpath = "/xyz/openbmc_project/inventory";
+    int depth = 0;
+    std::vector<std::string> pcieAdapterInterface = {
+        "xyz.openbmc_project.Inventory.Item.PCIeDevice"};
+    pldm::utils::GetSubTreeResponse response =
```

Comment:
can we use GetSubTreePaths here?

Reply:
No we cannot, we need the resposne the  GetSubTreePaths data type here